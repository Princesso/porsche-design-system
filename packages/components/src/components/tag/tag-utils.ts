import { TAG_DISMISSIBLE_COLORS } from '../tag-dismissible/tag-dismissible-utils';
import type { Theme } from '../../types';
import { isThemeDark } from '../../utils/theme/isThemeDark';
import type { ThemedColors } from '../../styles'; // deep import needed since barrel contains MutationObserver and causes VRT to fail because of TAG_COLORS import

export const TAG_COLORS = [
  ...TAG_DISMISSIBLE_COLORS,
  'neutral-contrast-high',
  'notification-neutral', // 'notification-neutral' is deprecated (replaced with 'notification-information')
  'notification-information',
  'notification-warning',
  'notification-success',
  'notification-error',
] as const;
export type TagColor = typeof TAG_COLORS[number];

export const hasInvertedThemeColor = (tagColor: TagColor, theme: Theme): boolean => {
  const isDark = isThemeDark(theme);
  return (
    (!isDark && tagColor === 'neutral-contrast-high') ||
    (isDark &&
      tagColor !== 'background-surface' &&
      tagColor !== 'background-default' &&
      tagColor !== 'notification-neutral' && // 'notification-neutral' is deprecated (replaced with 'notification-information')
      tagColor !== 'notification-information' &&
      tagColor !== 'notification-warning' &&
      tagColor !== 'notification-success' &&
      tagColor !== 'notification-error')
  );
};

export const getThemedBackgroundHoverColor = (tagColor: TagColor, themedColors: ThemedColors, theme: Theme): string => {
  const isDark = isThemeDark(theme);
  const colorMap: { [key in TagColor]: string } = {
    'background-default': isDark ? themedColors.backgroundColorLighten : themedColors.backgroundColorDarken,
    'background-surface': isDark
      ? themedColors.backgroundSurfaceColorLighten
      : themedColors.backgroundSurfaceColorDarken,
    'neutral-contrast-high': isDark ? themedColors.contrastHighColorLighten : themedColors.contrastHighColor,
    'notification-neutral': isDark ? themedColors.infoSoftColorLighten : themedColors.infoSoftColorDarken, // 'notification-neutral' is deprecated (replaced with 'notification-information')
    'notification-information': isDark ? themedColors.infoSoftColorLighten : themedColors.infoSoftColorDarken,
    'notification-success': isDark ? themedColors.successSoftColorLighten : themedColors.successSoftColorDarken,
    'notification-error': isDark ? themedColors.errorSoftColorLighten : themedColors.errorSoftColorDarken,
    'notification-warning': isDark ? themedColors.warningSoftColorLighten : themedColors.warningSoftColorDarken,
  };

  return colorMap[tagColor];
};
