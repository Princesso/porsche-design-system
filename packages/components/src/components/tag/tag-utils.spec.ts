import { hasInvertedThemeColor, TagColor } from './tag-utils';
import type { Theme } from '../../types';

describe('hasInvertedThemeColor()', () => {
  it.each<[TagColor, Theme, boolean]>([
    ['background-surface', 'light', false],
    ['background-default', 'light', false],
    ['neutral-contrast-high', 'light', true],
    ['notification-success', 'light', false],
    ['notification-warning', 'light', false],
    ['notification-error', 'light', false],
    ['notification-information', 'light', false],
    ['background-surface', 'dark', false],
    ['background-default', 'dark', false],
    ['neutral-contrast-high', 'dark', true],
    ['notification-success', 'dark', false],
    ['notification-warning', 'dark', false],
    ['notification-error', 'dark', false],
    ['notification-information', 'dark', false],
  ])('should for color: %s and theme: %s return %s', (color, theme, expected) => {
    expect(hasInvertedThemeColor(color, theme)).toBe(expected);
  });
});
