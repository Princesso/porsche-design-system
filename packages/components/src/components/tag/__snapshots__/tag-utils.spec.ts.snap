// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: background-default, theme: dark 1`] = `"#292934"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: background-default, theme: light 1`] = `"#E0E0E0"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: background-surface, theme: dark 1`] = `"#3E4045"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: background-surface, theme: light 1`] = `"#CBCED7"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: neutral-contrast-high, theme: dark 1`] = `"#CECFD1"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: neutral-contrast-high, theme: light 1`] = `"#535457"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-error, theme: dark 1`] = `"#331F1F"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-error, theme: light 1`] = `"#EED7D9"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-info, theme: dark 1`] = `"#0F2943"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-info, theme: light 1`] = `"#D6DEF0"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-neutral, theme: dark 1`] = `"#0F2943"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-neutral, theme: light 1`] = `"#D6DEF0"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-success, theme: dark 1`] = `"#0F432F"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-success, theme: light 1`] = `"#DDEAE1"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-warning, theme: dark 1`] = `"#362F1C"`;

exports[`getThemedBackgroundHoverColor() should return correct backgroundColor for color: notification-warning, theme: light 1`] = `"#F5ECD2"`;
